# Bengaluru-House-Price-Prediction

In [71]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline
import matplotlib
matplotlib.rcParams["figure.figsize"] = (20,20)

In [72]:
df1 = pd.read_csv("bengaluru_house_prices.csv")
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [73]:
#shows number of rows and columns
df1.shape

(13320, 9)

In [74]:
# The groupby function in Pandas is a tool that helps you organize data into groups 
# based on certain criteria, like the values in a column.

#the below line means, groupby area_type and count area_type
result = df1.groupby('area_type')['area_type'].agg('count')
display(result)

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

In [75]:
# Removing some columns, that are not necessary okey

# Drop specified labels from rows or columns (axis 0 means rows, and axis = 1 means columns)
df2 = df1.drop(['area_type', 'society', 'balcony', 'availability'], axis=1)
df2.head()


location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [76]:
# Data Cleaning
# checking the number of NA/Null examples

# The output of df.isnull().sum() is a Pandas Series
# where the index represents the column names of the
# original DataFrame and the values represent the count of
# null values in each respective column
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [77]:
# Dropping off the null/NA examples

df3 = df2.dropna()
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [78]:
df3.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [79]:
# Since hamare pass 2bhk bhi hai and also 2 bedroom, both means same

df3['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

The apply() method in Pandas is a versatile function used to apply a custom or built-in function along an axis of a DataFrame or to each element of a Series. Its primary purpose is to enable flexible data transformations and manipulations that might not be directly available through vectorized Pandas operations.

 Lambda Function: Its primary work involves applying custom logic or transformations to data without the need for formally defining a separate function.

In [80]:
df3['bhk'] = df3['size'].apply(lambda x : int(x.split(" ")[0]))

C:\Users\Baba_\AppData\Local\Temp\ipykernel_23324\3176993600.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['bhk'] = df3['size'].apply(lambda x : int(x.split(" ")[0]))


In [81]:
df3.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [82]:
df3['bhk'].unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18])

How do we have a 43 bedroom(bhk)

In [83]:
df3[df3.bhk>20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

how come 43 bedroom is fitted in 2400sq_ft??? It's like a error

In [84]:
df3.total_sqft.unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      shape=(2067,), dtype=object)

1133 - 1384  we got range in total Sqft

i need in single digit number

In [85]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

# What happens is that in try block "x" a number tries to convert a value in float
# if it can be converted in float its good and returns true, but if its something like
# a range of value like in total_sqft 1133-1145 and all, it goes to except block and returns false

In [86]:
# we apply "~" at the start to look for values that have been 
# returned false in the "is_float" func

df3[~df3['total_sqft'].apply(is_float)].head(15)

location       size      total_sqft  bath    price  bhk
30               Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122                 Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137     8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165               Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188               KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410                Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549            Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648                Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661              Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672           Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4
772  Banashankari Stage VI      2 BHK     1160 - 1195   2.0   59.935    2
775           Basavanagara      1 BHK   1000Sq. Meter   2.0   93.000    1
850      Bannerghatta Road      2 BHK     1115 - 1130   2.0   58.935    2
872      Singapura Village      2 BHK   1100Sq. Yards   2.0   45.000    2
886             Chandapura      1 BHK       520 - 645   1.0   15.135    1

we get values in total_sqft like perch, Sq. Meter, and ranges, and all like unstructured and stuffs, not clean

In [87]:
#  Cleaning values like these
def convert_sqft_to_num(x):
    token = x.split('-')
    if len(token) == 2:
        range_sum = float(token[0])+float(token[1])
        avg = range_sum/2
        return avg
    

    try:
        return float(x)
    except:
        return None

In [88]:
# Sample Run
convert_sqft_to_num('3156Sq.mt')

In [89]:
# .copy() does a deep copy of original dataframe
df4 = df3.copy()
df4['total_sqft'] = df4['total_sqft'].apply(lambda x:convert_sqft_to_num(x))
df4.head(15)

location       size  total_sqft  bath   price  bhk
0   Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1           Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2                Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3         Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                   Kothanur      2 BHK      1200.0   2.0   51.00    2
5                 Whitefield      2 BHK      1170.0   2.0   38.00    2
6           Old Airport Road      4 BHK      2732.0   4.0  204.00    4
7               Rajaji Nagar      4 BHK      3300.0   4.0  600.00    4
8               Marathahalli      3 BHK      1310.0   3.0   63.25    3
9               Gandhi Bazar  6 Bedroom      1020.0   6.0  370.00    6
10                Whitefield      3 BHK      1800.0   2.0   70.00    3
11                Whitefield  4 Bedroom      2785.0   5.0  295.00    4
12        7th Phase JP Nagar      2 BHK      1000.0   2.0   38.00    2
13                 Gottigere      2 BHK      1100.0   2.0   40.00    2
14                  Sarjapur  3 Bedroom      2250.0   3.0  148.00    3

checking if value at index 30 which was before an range of sqft number 

In [90]:
df4.loc[30]

# Yes, its now a single valued

location      Yelahanka
size              4 BHK
total_sqft       2475.0
bath                4.0
price             186.0
bhk                   4
Name: 30, dtype: object